### PR TITLE
[#1523] MQTT adapter: Let gateways subscribe for commands of a specific device

### PR DESCRIPTION
**Pull request by calohmn**
This is for #1523.

### CODE REVIEW

**Comment by calohmn** [comment]
Note: When a gateway has subscribed for commands _of a specific device_ (as introduced as a feature here), the topic, that the gateway will receive the commands on, will still contain the device id, although that is in a way superfluous.

Nonetheless, I think it's better to use a common topic structure along with the case that the gateway has subscribed for commands of _all_ his connected devices.

Relevant code fragment:
https://github.com/eclipse/hono/blob/90a391f100be39b2e500ade589e73d071d306ac1/adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java#L1356-L1364

File: adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java [hunk 1]
```diff
@@ -730,47 +731,51 @@ protected final void onUnsubscribe(
...
+                        sub.getTenant(),
+                        tenantTracker.result(),
+                        ProcessingOutcome.from(failure),
+                        command.getPayloadSize(),
+                        timer);
+                return null;
+            });
+        };
+        final Handler<Void> remoteCloseHandler = remoteClose -> {};
+        if (sub.getAuthenticatedDeviceId() != null && !sub.getAuthenticatedDeviceId().equals(sub.getDeviceId())) {
```
Comment (sophokles73) [inline adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java]: this check looks like it should be implemented in a method (e.g. *isGatewaySubscription*) in `Subscription`, don't you think?

File: adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/CommandSubscription.java [hunk 2]
```diff
@@ -101,14 +102,14 @@ private CommandSubscription(final String topic, final Device authenticatedDevice
             }
         } else {
             isAuthenticated = true;
-            if ((tenant != null && !authenticatedDevice.getTenantId().equals(tenant)) ||
-                    (deviceId != null && !authenticatedDevice.getDeviceId().equals(deviceId))) {
+            if (tenant != null && !authenticatedDevice.getTenantId().equals(tenant)) {
```
Comment (sophokles73) [inline adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/CommandSubscription.java]: I think the message in the thrown exception needs to be adapted as well

File: adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java (hunk 1 above)
Comment (calohmn) [reply adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java]: Good point, I've refactored it accordingly.

File: adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/CommandSubscription.java (hunk 2 above)
Comment (calohmn) [reply adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/CommandSubscription.java]: Yes, I've adapted it.

**Pull request merged**